Python/immutability.c

```diff
@@ -0,0 +1,403 @@
+    }else{
+        module_dict = NULL;
+    }
+
+    _Py_VISIT_FUNC_ATTR(f->func_defaults, frontier);
+    _Py_VISIT_FUNC_ATTR(f->func_kwdefaults, frontier);
+    _Py_VISIT_FUNC_ATTR(f->func_doc, frontier);
+    _Py_VISIT_FUNC_ATTR(f->func_name, frontier);
+    _Py_VISIT_FUNC_ATTR(f->func_dict, frontier);
+    _Py_VISIT_FUNC_ATTR(f->func_closure, frontier);
+    _Py_VISIT_FUNC_ATTR(f->func_annotations, frontier);
+    _Py_VISIT_FUNC_ATTR(f->func_typeparams, frontier);
+    _Py_VISIT_FUNC_ATTR(f->func_qualname, frontier);
+
+    f_stack = PyList_New(0);
+    if(f_stack == NULL){
+        return PyErr_NoMemory();
+    }
+
+    f_ptr = f->func_code;
+    if(push(f_stack, f_ptr)){
+        goto nomemory;
+    }
+
+    frozen_builtins = PyDict_New();
+    if(frozen_builtins == NULL){
+        goto nomemory;
+    }
+
+    frozen_globals = PyDict_New();
+    if(frozen_globals == NULL){
+        goto nomemory;
+    }
+
+    while(PyList_Size(f_stack) != 0){
+        f_ptr = pop(f_stack);
+        _PyObject_ASSERT(f_ptr, PyCode_Check(f_ptr));
+        f_code = (PyCodeObject*)f_ptr;
+
+        size = 0;
+        if (f_code->co_names != NULL)
+          size = PySequence_Fast_GET_SIZE(f_code->co_names);
+        for(Py_ssize_t i = 0; i < size; i++){
+            PyObject* name = PySequence_Fast_GET_ITEM(f_code->co_names, i);
+
+            if(PyUnicode_CompareWithASCIIString(name, "globals") == 0){
+                // if the code calls the globals() builtin, then any
+                // cellvar or const in the function could, potentially, refer to
+                // a global variable. As such, we need to check if the globals
+                // dictionary contains that key and then make it immutable
+                // from this point forwards.
+                check_globals = true;
+            }
+
+            if(PyDict_Contains(globals, name)){
+                PyObject* value = PyDict_GetItem(globals, name);
+                if(PyDict_SetItem(frozen_globals, name, value)){
+                    Py_DECREF(frozen_builtins);
+                    Py_DECREF(frozen_globals);
+                    Py_DECREF(f_stack);
+                    return NULL;
+                }
+            }else if(PyDict_Contains(builtins, name)){
+                PyObject* value = PyDict_GetItem(builtins, name);
+                if(PyDict_SetItem(frozen_builtins, name, value)){
+                    Py_DECREF(frozen_builtins);
+                    Py_DECREF(frozen_globals);
+                    Py_DECREF(f_stack);
+                    return NULL;
+                }
+            }else if(PyDict_Contains(module_dict, name)){
+                PyObject* value = PyDict_GetItem(module_dict, name);
+
+                _PyDict_SetKeyImmutable((PyDictObject*)module_dict, name);
+
+                if(!_Py_IsImmutable(value)){
+                    if(push(frontier, value)){
+                        goto nomemory;
+                    }
+                }
+            }
```

Comment:
I couldn't work out how to exercise this code.  It doesn't appear to be hit by `test_freeze`.  Should we be able to get coverage for this?  I had an experiment with the REPL and a few modules, but this code path didn't get exercised.